### PR TITLE
Install providers

### PR DESCRIPTION
**Pull request by asalkeld**
```
oc get all,cm
NAME                                                    READY   STATUS    RESTARTS   AGE
pod/capi-controller-manager-6c8b76f9f4-tb6gw            1/1     Running   0          3h6m
pod/capi-operator-controller-manager-6bb45454c9-lzmc2   2/2     Running   0          14m
pod/capm3-controller-manager-656795c95-xfjgd            1/1     Running   0          3h6m
pod/cluster-capi-operator-5b95fdd459-8fkm9              1/1     Running   0          17m

NAME                                                       TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
service/capi-operator-controller-manager-metrics-service   ClusterIP   172.30.63.84     <none>        8443/TCP   22h
service/capi-webhook-service                               ClusterIP   172.30.35.109    <none>        443/TCP    3h6m
service/capm3-webhook-service                              ClusterIP   172.30.160.121   <none>        443/TCP    3h6m
service/ipam-webhook-service                               ClusterIP   172.30.6.86      <none>        443/TCP    3h6m

NAME                                               READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/capi-controller-manager            1/1     1            1           3h6m
deployment.apps/capi-operator-controller-manager   1/1     1            1           14m
deployment.apps/capm3-controller-manager           1/1     1            1           3h6m
deployment.apps/cluster-capi-operator              1/1     1            1           17m

NAME                                                          DESIRED   CURRENT   READY   AGE
replicaset.apps/capi-controller-manager-6c8b76f9f4            1         1         1       3h6m
replicaset.apps/capi-operator-controller-manager-6bb45454c9   1         1         1       14m
replicaset.apps/capm3-controller-manager-656795c95            1         1         1       3h6m
replicaset.apps/cluster-capi-operator-5b95fdd459              1         1         1       17m

NAME                                                 DATA   AGE
configmap/aws-v0.7.0                                 2      5h44m
configmap/azure-v0.5.2                               2      5h44m
configmap/cluster-api-v1.0.0                         2      5h44m
configmap/cluster-capi-operator-images               1      12d
configmap/cluster-capi-operator-leader               0      12d
configmap/controller-leader-election-capi-operator   0      12d
configmap/gcp-v0.4.0                                 2      5h44m
configmap/kube-root-ca.crt                           1      12d
configmap/metal3-v0.5.1                              2      5h44m
configmap/openshift-service-ca.crt                   1      12d
configmap/openstack-v0.4.0                           2      5h44m

```

### CODE REVIEW

**Review by alexander-demichev — APPROVED** [review]
/approve 
I have couple comments but they can be addressed in follow-up PRs:

- We should add code that prevents reconciliation when provider is not supported for tech preview.
- Place provider and operator installation in separate files, I think we might end up with a lot of code placed inside one file. 
- Should we remove BootstrapProvider and ControlPlaneProvider mutation logic from the controller for now? 
- After we set up all forks we can use manifests from forks instead of upstream.

**Comment by asalkeld** [comment]
I have just added an option to the upsteam operator "--delete-rbac-on-upgrade=true/false" (default true). 
in https://github.com/openshift/cluster-capi-operator/pull/8/commits/d32b707bf7850b9939bc374ab5284c5ab25d494f I have
set this to false. It depends on: https://github.com/asalkeld/cluster-api/commit/89c26edb76851df9f064ac645d49e6438d69776c

Why do this? Upstream has not very smart resource update that will fail if rolebindings have a role reference change.
also in the spec it says to delete resources when the version of the provider is increased.
However we are now installing rbac resources using the /manifests/ and we do not have permission (or the desire) to
delete these rbac resources.

**Comment by asalkeld** [comment]
>  I have couple comments but they can be addressed in follow-up PRs:
> 
>     * We should add code that prevents reconciliation when provider is not supported for tech preview.

This is done in the Reconcile, but yes it could be done in featureGatePredicates()

> 
>     * Place provider and operator installation in separate files, I think we might end up with a lot of code placed inside one file.

capi-operator related assets are in /assets/capi-operator/*.yaml
provider related assets are in /assets/providers/*.yaml 

> 
>     * Should we remove BootstrapProvider and ControlPlaneProvider mutation logic from the controller for now?

Done

> 
>     * After we set up all forks we can use manifests from forks instead of upstream.

Sure,should be do-able.

**Review by elmiko — COMMENTED** [review]
i'm generally +1 to merging this, we are holding a lazy consensus until end of day on tuesday 16 nov

**Comment by asalkeld** [comment]
> i'm generally +1 to merging this, we are holding a lazy consensus until end of day on tuesday 16 nov

Cool, I do have some docs/hacking.md changes on the way and as-yet-untested cluster.status reconciler - but would be good to get this in.

**Review by JoelSpeed — COMMENTED** [review]
I've tried to get through some of this, but there's a lot, so some hasn't been thoroughly reviewed by me at least (Eg RBAC files), but for the most part I think this looks good, have a few questions/comments I'd appreciate if we could quickly discuss before we merge

File: .gitleaks.toml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[allowlist]
+  files = [ '''assets/bindata.go''']
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply .gitleaks.toml:2]: oops, i'll get rid of this

File: go.mod [hunk 2]
```diff
@@ -4,21 +4,25 @@ go 1.16
...
+	sigs.k8s.io/cluster-api v0.4.3 // indirect
+	sigs.k8s.io/cluster-api/exp/operator v0.0.0-00010101000000-000000000000
+	sigs.k8s.io/controller-runtime v0.10.1
 )
 
-replace github.com/go-logr/logr => github.com/go-logr/logr v0.4.0
+replace (
+	github.com/go-logr/logr => github.com/go-logr/logr v0.4.0
+	sigs.k8s.io/cluster-api => github.com/asalkeld/cluster-api v0.4.1-0.20210923065712-6ed39b7ef8f9
+	sigs.k8s.io/cluster-api/exp/operator => github.com/asalkeld/cluster-api/exp/operator v0.0.0-20211005030408-a8791e47e147
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply go.mod:27]: I'd love to, but that is where it is right now. The upstream repo has not been made yet..

[the remark this answers is not in the document]
Comment (asalkeld) [reply go.mod:27]: Note: this is for the CoreProvider and InfrastructureProvider types defined by the upstream operator.

[the remark this answers is not in the document]
Comment (JoelSpeed) [reply go.mod:27]: Ack, that makes sense, let's merge and clean up later then

**Pull request merged**